Implement static start method for HelpDoclet

magicDGS opened the pull request:
This is required to allow the usage of the HelpDoclet without extension
of the class. Includes an update of the javadoc to clarify that the
method is required.

Close #55

magicDGS commented:
Small bug and documentation fix for make `HelpDoclet` usable, @cmnbroad. Can you review, please?

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java:
```diff
@@ -82,6 +84,14 @@
 
     /**
+     * The entry point to javadoc generation. It calls {@link #startProcessDocs(RootDoc)} (String[])} from this instance.
+     * Consumers should implement this static method to call {@link #startProcessDocs(RootDoc)} from their instance.
```
Since this method is static, it probably shouldn't mention "this instance". Can you reword this to something like "The entry point for javadoc generation. It creates an instance of HelpDoclet and calls startProcessDocs on that instance. Custom Barclay doclets should subclass this class, and implement a similar static method that creates an instance of the doclet subclass and delegates to that instance's startProcessDocs".

magicDGS replied:
Done with small modifications.

magicDGS commented:
I will add a test, but for that I will also need to modify the help templates and copy the current ones to the test resources for the test doclet. Do you mind if I do this, @cmnbroad? The custom templates are not working because they mention the 'supercat' tag and maybe others (I haven't look into detail). I do not want to go too much into this (better to design this while addressing #39), but having non-specific template will help with development of custom template.

My idea behind these changes is to have working the custom doclet to make easy to have a quick check if the classes are working with the default one and the default templates. Then, if the user need it, they could add custom features and make it more complicated.

Waiting for your answer here, @cmnbroad!

cmnbroad commented:
@magicDGS Yeah, just a simple test using simple templates should do. It will ensure that we don't break this use case. Preferably put them in a separate folder (you might want to move the existing ones so the two sets can live in parallel folders that correspond to the tests).

magicDGS commented:
I'll do it.

My idea was more related to change the structure of resources by the following:

* Move the current templates to the `test/resources/testTemplates` as they are
* Derive new default templates from them, simplifying specific-features tested in the current implementation.
* These templates could be used as started point for custom templates (as it is stated in the current Wiki). They will leave in `main/resources/org/broadinstitute/barclay/helpTemplates`, allowing to pack them into the jar file and use them as-is in projects that requires a very simple documentation template.

What do you think about it, @cmnbroad? I can do this here or in a different ticket, but I believe that this organization may benefit other user with less requirements....

cmnbroad commented:
@magicDGS That sounds like a pretty good plan. Thanks.

magicDGS commented:
I repackaged the templates, removed in the supercat in the default ones, and added a test for the default doclet.

Thanks for reviewing - back to you, @cmnbroad!

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java:
```diff
@@ -41,6 +41,8 @@
  * {@link #createGSONWorkUnit}
  *
+ * This requires that the static {@link #start(RootDoc)} methods is implemented to call the
+ * {@link #startProcessDocs(RootDoc)} from their instance.
```
I would remove the blank line (43) so this appears with the list right above, and change it to say:

{@link #start(RootDoc)} A static method that instantiates the subclass and delegates to the instance method {@link #startProcessDocs(RootDoc)}.

magicDGS replied:
Done.

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java:
```diff
@@ -65,4 +65,44 @@ public boolean filesContentsIdentical(
+                "-settings-dir", "src/main/resources/org/broadinstitute/barclay/helpTemplates/",
+                "-d", outputDir.getAbsolutePath(),
+                "-output-file-extension", "html",
+                "-doclet", TestDoclet.class.getName(),
```
A rebase problem, sorry. I change it!

magicDGS commented on src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java:
```diff
@@ -65,4 +65,44 @@ public boolean filesContentsIdentical(
     }
 
+    @Test
+    public void testDefaultDoclet() throws IOException {
```
Will do in second commit.

magicDGS replied:
Done.

magicDGS commented:
I addressed your comments. Back to you @cmnbroad!

cmnbroad commented:
@magicDGS Looks good - thanks for doing all that!

magicDGS commented:
Thanks!
